src/java/org/apache/cassandra/db/memtable/TrieMemtable.java

```diff
@@ -0,0 +1,408 @@
+            columnsCollector.update(update.columns());
+            statsCollector.update(update.stats());
+            currentOperations.addAndGet(update.operationCount());
+        }
+
+        if (partitions.reachedAllocatedSizeThreshold() && !switchRequested.getAndSet(true))
+        {
+            logger.info("Scheduling flush due to trie size limit reached.");
+            owner.signalFlushRequired(this, ColumnFamilyStore.FlushReason.MEMTABLE_LIMIT);
+        }
+
+        return updater.colUpdateTimeDelta;
+    }
+
+    public int partitionCount()
+    {
+        return partitions.valuesCount();
+    }
+
+    public MemtableUnfilteredPartitionIterator makePartitionIterator(final ColumnFilter columnFilter, final DataRange dataRange)
+    {
+        AbstractBounds<PartitionPosition> keyRange = dataRange.keyRange();
+
+        PartitionPosition left = keyRange.left;
+        PartitionPosition right = keyRange.right;
+        if (left.isMinimum())
+            left = null;
+        if (right.isMinimum())
+            right = null;
+
+        boolean isBound = keyRange instanceof Bounds;
+        boolean includeStart = isBound || keyRange instanceof IncludingExcludingBounds;
+        boolean includeStop = isBound || keyRange instanceof Range;
+
+        Trie<BTreePartitionData> subMap = partitions.subtrie(left, includeStart, right, includeStop);
+
+        return new MemtableUnfilteredPartitionIterator(metadata(),
+                                                       allocator.ensureOnHeap(),
+                                                       subMap,
+                                                       columnFilter,
+                                                       dataRange);
+    }
+
+    public Partition getPartition(DecoratedKey key)
+    {
+        BTreePartitionData data = partitions.get(key);
+        if (data != null)
+            return createPartition(metadata(), allocator.ensureOnHeap(), key, data);
+        else
+            return null;
+    }
+
+    private static MemtablePartition createPartition(TableMetadata metadata, EnsureOnHeap ensureOnHeap, DecoratedKey key, BTreePartitionData data)
+    {
+        return new MemtablePartition(metadata, ensureOnHeap, key, data);
+    }
+
+    private static MemtablePartition getPartitionFromTrieEntry(TableMetadata metadata, EnsureOnHeap ensureOnHeap, Map.Entry<ByteComparable, BTreePartitionData> en)
+    {
+        DecoratedKey key = BufferDecoratedKey.fromByteComparable(en.getKey(),
+                                                                 BYTE_COMPARABLE_VERSION,
+                                                                 metadata.partitioner);
+        return createPartition(metadata, ensureOnHeap, key, en.getValue());
+    }
+
+
+    public FlushCollection<MemtablePartition> getFlushSet(PartitionPosition from, PartitionPosition to)
+    {
+        Trie<BTreePartitionData> toFlush = partitions.subtrie(from, true, to, false);
```

Comment:
could we assert that the memtable has been switched out? Same for other implementations?

One thing to consider is to whether ensure this contract on the level of API (e.g. a non-virtual interface for `getFlushSet` or `discard` which does the verification + overriden methods that do the job, like `doDiscard` etc). I'm not a fan of such solution, but perhaps there is a reason to do it that way.

Reply:
No. Parts of the memtable can be flushed without it being switched out (e.g. for streaming from memtable).